producer/buffer/buffer.go

```diff
@@ -0,0 +1,215 @@
+// THE SOFTWARE.
+
+package buffer
+
+import (
+	"container/list"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3msg/producer"
+	"github.com/m3db/m3msg/producer/data"
+
+	"github.com/uber-go/atomic"
+	"github.com/uber-go/tally"
+)
+
+var (
+	errBufferFull                         = errors.New("buffer full")
+	errBufferClosed                       = errors.New("buffer closed")
+	errInvalidDataLargerThanMaxBufferSize = errors.New("invalid data, larger than max buffer size")
+)
+
+type bufferMetrics struct {
+	messageDropped  tally.Counter
+	bytesDropped    tally.Counter
+	dataTooLarge    tally.Counter
+	messageBuffered tally.Gauge
+	bytesBuffered   tally.Gauge
+}
+
+func newBufferMetrics(scope tally.Scope) bufferMetrics {
+	return bufferMetrics{
+		messageDropped:  scope.Counter("message-dropped"),
+		bytesDropped:    scope.Counter("bytes-dropped"),
+		dataTooLarge:    scope.Counter("data-too-large"),
+		messageBuffered: scope.Gauge("message-buffered"),
+		bytesBuffered:   scope.Gauge("bytes-buffered"),
+	}
+}
+
+type buffer struct {
+	sync.RWMutex
+
+	buffers       *list.List
+	opts          Options
+	maxBufferSize uint64
+	onFinalizeFn  data.OnFinalizeFn
+	m             bufferMetrics
+
+	size     *atomic.Uint64
+	doneCh   chan struct{}
+	isClosed bool
+}
+
+// NewBuffer returns a new buffer.
+func NewBuffer(opts Options) producer.Buffer {
+	if opts == nil {
+		opts = NewBufferOptions()
+	}
+	b := &buffer{
+		buffers:       list.New(),
+		maxBufferSize: uint64(opts.MaxBufferSize()),
+		opts:          opts,
+		m:             newBufferMetrics(opts.InstrumentOptions().MetricsScope()),
+		size:          atomic.NewUint64(0),
+		doneCh:        make(chan struct{}),
+		isClosed:      false,
+	}
+	b.onFinalizeFn = b.subSize
+	return b
+}
+
+func (b *buffer) Buffer(d producer.Data) (producer.RefCountedData, error) {
+	b.Lock()
+	if b.isClosed {
+		b.Unlock()
+		return nil, errBufferClosed
+	}
+	var (
+		dataSize      = uint64(d.Size())
+		maxBufferSize = b.maxBufferSize
+	)
+	if dataSize > maxBufferSize {
+		b.Unlock()
+		b.m.dataTooLarge.Inc(1)
+		return nil, errInvalidDataLargerThanMaxBufferSize
+	}
+	targetBufferSize := maxBufferSize - dataSize
+	if b.size.Load() > targetBufferSize {
+		if err := b.produceOnFullWithLock(targetBufferSize); err != nil {
+			b.Unlock()
+			return nil, err
+		}
+	}
+	b.size.Add(dataSize)
+	rd := data.NewRefCountedData(d, b.onFinalizeFn)
+	b.buffers.PushBack(rd)
+	b.Unlock()
+	return rd, nil
+}
+
+func (b *buffer) produceOnFullWithLock(targetSize uint64) error {
+	switch b.opts.OnFullStrategy() {
+	case ReturnError:
+		return errBufferFull
+	case DropEarliest:
+		b.dropEarliestUntilTargetWithLock(targetSize)
+	}
+	return nil
+}
+
+func (b *buffer) dropEarliestUntilTargetWithLock(targetSize uint64) {
+	var next *list.Element
+	for e := b.buffers.Front(); e != nil && b.size.Load() > targetSize; e = next {
+		next = e.Next()
+		d := e.Value.(producer.RefCountedData)
+		if !d.IsDroppedOrConsumed() {
```

Comment:
Can this run into a race condition, where right after you check the condition, the data gets consumed? If so, would that be an issue?

Reply:
Sounds good.